website/identifiers/utils.py

```diff
@@ -56,83 +55,73 @@ def merge_dicts(*dicts):
     return dict(sum((each.items() for each in dicts), []))
 
 
-def get_doi_and_metadata_for_object(target_object):
-    from osf.models import PreprintService
-
-    metadata_function = datacite_metadata_for_node
-    if isinstance(target_object, PreprintService):
-        metadata_function = datacite_metadata_for_preprint
+def get_doi_client(target_object):
+    """ Get the approprite DOI creation client for the target object requested.
+    :param target_object: object to request a DOI for.
+    :return: client appropriate for that target object.
+             If credentials for that target object aren't set, return None
+    """
+    from website.identifiers.clients import DataCiteClient, CrossRefClient
+    from osf.models import PreprintService, AbstractNode
 
-    doi = settings.EZID_FORMAT.format(namespace=settings.DOI_NAMESPACE, guid=target_object._id)
-    datacite_metadata = metadata_function(target_object, doi)
+    if isinstance(target_object, PreprintService) and settings.CROSSREF_DEPOSIT_URL:
```

Comment:
Type-checking this way is a bit smelly. It seems like subclasses of `IdentifierMixin` should be responsible for getting their own client.

```python
class IdentifierMixin:

    def get_doi_client(self):
        """Return a BaseIdentifierClient if proper
        settings are configured, else return None.
        """
        raise NotImplementedError



class AbstractNode(IdentifierMixin):

    def get_doi_client(self):
        if settings.DATACITE_URL:
            return DataciteClient()
        else:
            return None
```